### PR TITLE
Widget tab

### PR DESCRIPTION
**Pull request by Hex27**
- Added a Widget Tab
- Widgets now each have an editable title
- Added React.memo() around WidgetComp, and forced WidgetComp react keys to be the widget object's uuid. This reduced refreshes in some cases, but not all (i.e. deleting a widget)

### CODE REVIEW

**Comment by Hex27** [comment]
Don't merge, there's a lag issue now

File: devfiles/src/classes/Panel.ts [hunk 1]
```diff
@@ -162,14 +162,14 @@ export default class Panel {
...
         this.dataFilterer.dataUpdated();
-        this.refreshWidgets();
 
         //Update options to reflect new filters
         this.updateInputOptions();
         this.pageManager.refreshPanelOptions();
 
         //Refresh after internal class state is updated
         this.refreshComponent();
+        this.refreshWidgets();
```
Comment (CGDogan) [inline devfiles/src/classes/Panel.ts:172]: @SatoruQ You had moved it up, would moving it back down cause any problems for your code?

File: devfiles/src/classes/data/DataFilterer.ts [hunk 2]
```diff
@@ -55,6 +55,7 @@ export default class DataFilterer {
     // columnIndex can't be larger than the number of total columns
     // filterClass needs to be correct type, won't be checked here
     public replaceFilter(columnIndex, filterClass: Filter) {
+        this.filterState = uuidv4(); //Filters changed
```
Comment (CGDogan) [inline devfiles/src/classes/data/DataFilterer.ts:58]: Are you sure you don't want to add this to dataUpdated()?

File: devfiles/src/classes/Panel.ts (hunk 1 above)
Comment (Hex27) [reply devfiles/src/classes/Panel.ts:172]: Apparently not, I tested his code and it was ok

File: devfiles/src/classes/data/DataFilterer.ts (hunk 2 above)
Comment (Hex27) [reply devfiles/src/classes/data/DataFilterer.ts:58]: dataUpdated has this.dataState = uuidv4

**Review by CGDogan — APPROVED** [review]

**Pull request merged**